### PR TITLE
Redesign of MSA retirement to use signed extension

### LINKED ISSUE #44: As a user, I should be able to retire my MSA
As a current user of Frequency, I want to be able to remove myself from Frequency so that I can remove my association with Frequency.

# Description
Users should be able retire an MSA so that it cannot be used again.

## Notes
- Garbage collection of messages will be addressed in a different story. #128
- Fee:No - this is handled by #340 
- First attempt: https://github.com/LibertyDSNP/frequency/pull/252

# Acceptance Criteria
- [x] The user should be able to call the extrinsic that retires an MSA from the polkadot js ui
- [x] If the delegator does not have a MSA yet, the extrinsic should fail with `NoKeyExists`
- [x] All delegations should be revoked when calling the RPC to get delegations by MSA
- [x] I should not be able to call this unless there is only one key reference in `MsaInfoOf` and that is the key I am using to call from.
- [x] An event should be emitted to notify about the retirement Suggested: `MSARetired(msa_id)`
- [x] The `KeyRemoved` event should be triggered for the last key
- [x] I should not be able to call this if the MSA is a registered provider
- [ ] Benchmarks
- [x] Unit tests
- [x] New transactions associated with the MSA must be rejected
  - [x] FAIL: Add new key
  - [x] FAIL: Add new delegation
  - [x] FAIL: Modify delegation permissions
  - [x] FAIL: Add Message

## State Maps for Deletion
- [x] MSA -> u8 (references) deleted `MsaInfoOf`
- [x] Key -> MSA (Only one will be left) `MessageSourceIdOf`
- [x] MSA -> Provider deleted `ProviderInfoOf` (Should have none left)

## Work that doesn't need to be done
- `ProviderInfoOf` should already be emptied
- No new errors, as once `MessageSourceIdOf` deletion happens `NoKeyExists` should be triggered already.
- Pays::No is in #340 
- Garbage collection of messages will be addressed in a different story. #128

## Arch Notes
- Retire MSA extrinsic Signature: `retire_msa(origin: OriginFor<T>) -> DispatchResult`

## Other Notes
- Does not need DispatchResultWithPostInfo as we enforce that all the keys but one and all delegations are revoked.

### PR DESCRIPTION
**Pull request by rlaferla**
Implemented redesign of MSA retirement to use signed extension Implemented unit tests for MSA retirement

Closes #44

### CODE REVIEW

File: pallets/msa/src/lib.rs [hunk 1]
```diff
@@ -646,33 +646,29 @@ pub mod pallet {
 
 		/// Retire a MSA
 		///
+		/// When a user wants to disassociate themselves from Frequency, they can retire their MSA for free provided that:
+		///  (1) They own the MSA
+		///  (2) There is only one account key
+		///  (3) The MSA is not a registered provider.
+		///
+		/// This does not currently remove any messages related to the MSA.
```
Comment (saraswatpuneet) [inline pallets/msa/src/lib.rs:654]: Can we add a documentation for #Arguments that this extrinsic takes, you can check other examples of adding arguments like ``add_key_to_msa```

**Review by saraswatpuneet — APPROVED** [review]

**Review by enddynayn — APPROVED** [review]
Fantastic! 👍

File: pallets/msa/src/lib.rs (hunk 1 above)
Comment (rlaferla) [reply pallets/msa/src/lib.rs:654]: Definitely.  I see many other functions that are missing documentation for #Arguments as well.  We should document all of them but I'll add them for this one function.

Comment (rlaferla) [reply pallets/msa/src/lib.rs:654]: So for the `retire_msa()` function, it just takes in the `origin` as an argument.  I will document it.

**Pull request merged**